src/rust/src/rustzcash.rs

```diff
@@ -259,55 +199,45 @@ pub extern "C" fn librustzcash_merkle_hash(
     result: *mut [c_uchar; 32],
 ) {
     // Should be okay, because caller is responsible for ensuring
-    // the pointer is a valid pointer to 32 bytes, and that is the
-    // size of the representation
-    let a_repr = read_fr(unsafe { &*a });
-
-    // Should be okay, because caller is responsible for ensuring
-    // the pointer is a valid pointer to 32 bytes, and that is the
-    // size of the representation
-    let b_repr = read_fr(unsafe { &*b });
-
-    let tmp = merkle_hash(depth, &a_repr, &b_repr);
+    // the pointers are valid pointers to 32 bytes.
+    let tmp = merkle_hash(depth, unsafe { &*a }, unsafe { &*b });
 
     // Should be okay, caller is responsible for ensuring the pointer
     // is a valid pointer to 32 bytes that can be mutated.
     let result = unsafe { &mut *result };
-    tmp.write_le(&mut result[..]).expect("length is 32 bytes");
+    *result = tmp;
 }
 
 #[no_mangle] // ToScalar
 pub extern "C" fn librustzcash_to_scalar(input: *const [c_uchar; 64], result: *mut [c_uchar; 32]) {
     // Should be okay, because caller is responsible for ensuring
     // the pointer is a valid pointer to 32 bytes, and that is the
     // size of the representation
-    let scalar = <Bls12 as JubjubEngine>::Fs::to_uniform(unsafe { &(&*input)[..] }).into_repr();
+    let scalar = jubjub::Scalar::from_bytes_wide(unsafe { &*input });
 
     let result = unsafe { &mut *result };
 
-    scalar
-        .write_le(&mut result[..])
-        .expect("length is 32 bytes");
+    *result = scalar.to_bytes();
```

Comment:
Not being familiar with this syntax before now, I went to investigate and found this:

> Storing through a raw pointer using *ptr = data calls drop on the old value, so write must be used if the type has drop glue and memory is not already initialized - otherwise drop would be called on the uninitialized memory.

When it says "the old value" does this refer to the previous value at the memory location previously pointed to by `*result`? This memory is initialized in the C++ code, but I'm not sure I understand what dropping that means.

Reply:
`result` is not a raw pointer at this point (🥁🥁🔔). On the previous line, we convert `result` into an `&mut [u8; 32]`; here we are dereferencing the mutable reference in order to assign it.

Reply:
Ah, okay, I looked right past the fact that `result` is shadowed.

Reply:
If `result` _were_ still a raw pointer, then yes, "the old value" would mean the C++ memory (which in our case will not be uninitialized because it's a `uint256` which zeroes itself on construction). There is no `Drop` implementation for plain arrays, so nothing would happen here in any case.